File: cmd/nogodey/logger/logger.go
package logger

import (
	"encoding/json"
	"fmt"
	"os"
	"time"
)

// Logger provides structured JSON logging
type Logger struct{}

// LogEntry represents a structured log entry
type LogEntry struct {
	Level     string                 `json:"level"`
	Message   string                 `json:"msg"`
	Timestamp string                 `json:"timestamp"`
	Fields    map[string]interface{} `json:",inline,omitempty"`
}

// New creates a new Logger instance
func New() *Logger {
	return &Logger{}
}

// Info logs an info level message with optional key-value metadata
func (l *Logger) Info(msg string, fields ...interface{}) {
	l.log("info", msg, fields...)
}

// Warn logs a warning level message with optional key-value metadata
func (l *Logger) Warn(msg string, fields ...interface{}) {
	l.log("warn", msg, fields...)
}

// Error logs an error level message with optional key-value metadata
func (l *Logger) Error(msg string, fields ...interface{}) {
	l.log("error", msg, fields...)
}

// log is the internal logging method that formats and outputs JSON
func (l *Logger) log(level, msg string, fields ...interface{}) {
	entry := LogEntry{
		Level:     level,
		Message:   msg,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		Fields:    make(map[string]interface{}),
	}

	for i := 0; i < len(fields)-1; i += 2 {
		if key, ok := fields[i].(string); ok {
			entry.Fields[key] = fields[i+1]
		}
	}

	if data, err := json.Marshal(entry); err == nil {
		fmt.Fprintln(os.Stdout, string(data))
	}
}

// Timer provides duration measurement and logging
type Timer struct {
	name  string
	start time.Time
}

// StartTimer creates and starts a new timer with the given name
func StartTimer(name string) *Timer {
	return &Timer{
		name:  name,
		start: time.Now(),
	}
}

// Observe logs the duration since the timer was started
func (t *Timer) Observe() {
	duration := time.Since(t.start)
	durationMs := float64(duration.Nanoseconds()) / 1e6

	logger := New()
	logger.Info("timer completed",
		t.name+"_duration_ms", durationMs,
		"operation", t.name,
	)
}

// ObserveWithLogger logs the duration using the provided logger
func (t *Timer) ObserveWithLogger(logger *Logger) {
	duration := time.Since(t.start)
	durationMs := float64(duration.Nanoseconds()) / 1e6

	logger.Info("timer completed",
		t.name+"_duration_ms", durationMs,
		"operation", t.name,
	)
}

// Global logger instance for convenience
var defaultLogger = New()

// Package-level convenience functions
func Info(msg string, fields ...interface{}) {
	defaultLogger.Info(msg, fields...)
}

func Warn(msg string, fields ...interface{}) {
	defaultLogger.Warn(msg, fields...)
}

func Error(msg string, fields ...interface{}) {
	defaultLogger.Error(msg, fields...)
}

File: js/src/logger.ts
export type Meta = Record<string, any>;

interface LogEntry {
  level: string;
  message: string;
  timestamp: string;
  [key: string]: any;
}

/**
 * Creates a structured log entry and outputs it as JSON
 */
function log(level: string, meta: Meta, message: string): void {
  const entry: LogEntry = {
    level,
    message,
    timestamp: new Date().toISOString(),
    ...meta,
  };

  console.log(JSON.stringify(entry));
}

/**
 * Logs an info level message with optional metadata
 */
export function info(meta: Meta, message: string): void {
  log('info', meta, message);
}

/**
 * Logs a warning level message with optional metadata
 */
export function warn(meta: Meta, message: string): void {
  log('warn', meta, message);
}

/**
 * Logs an error level message with optional metadata
 */
export function error(meta: Meta, message: string): void {
  log('error', meta, message);
}

/**
 * Timer utility for measuring operation duration
 */
export class Timer {
  private readonly name: string;
  private readonly start: number;

  constructor(name: string) {
    this.name = name;
    this.start = Date.now();
  }

  /**
   * Log the duration since the timer was started
   */
  observe(): void {
    const duration = Date.now() - this.start;
    info(
      { 
        [`${this.name}_duration_ms`]: duration,
        operation: this.name,
      },
      'timer completed'
    );
  }
}

/**
 * Create and start a new timer
 */
export function startTimer(name: string): Timer {
  return new Timer(name);
}

/**
 * Convenience function for timing async operations
 */
export async function timeAsync<T>(
  name: string,
  operation: () => Promise<T>
): Promise<T> {
  const timer = startTimer(name);
  try {
    const result = await operation();
    timer.observe();
    return result;
  } catch (error) {
    timer.observe();
    throw error;
  }
} 